R/getInsertSizeDistFromBam.R

```diff
@@ -0,0 +1,108 @@
+#'   regions will be used. If \code{NULL} (the default), all alignments are
+#'   used.
+#' @param nmax \code{numeric(1)} specifying the maximal number of alignments to
+#'   read. If \code{NA} (the default), the alignments in \code{regions} (if
+#'   \code{regions} are not \code{NULL}) or in the bam file will be used.
+#' @param isizemax \code{numeric(1)} specifying the maximal insert size to
+#'   report. Larger insert sizes will be set to \code{isizemax} with a warning.
+#'
+#' @return \code{integer} vector with the number of insert sizes. The element at
+#'   position \code{i} gives the observed number of alignment pairs with an
+#'   insert size of \code{i}. The number of insert sizes greater than
+#'   \code{isizemax} that were set to \code{isizemax} are reported in the
+#'   attribute \code{"ncapped"}.
+#'
+#' @seealso \code{\link[Rsamtools]{scanBam}} used to read alignments.
+#'
+#' @examples
+#' if (requireNamespace("Rsamtools", quietly = TRUE)) {
+#'     bamf <- system.file("extdata", "getInsertSizeDistFromBam", "atac_mm10.bam",
+#'                         package = "swissknife")
+#'     isize <- getInsertSizeDistFromBam(bamf)
+#'     attr(isize, "ncapped")
+#'     plot(isize, type = "l",
+#'          xlab = "Insert size (bp)", ylab = "Number of fragments")
+#' }
+#'
+#' @importFrom IRanges IRangesList
+#' 
+#' @export
+getInsertSizeDistFromBam <- function(fname,
+                                     regions = NULL,
+                                     nmax = NA_integer_,
+                                     isizemax = 800) {
+    .assertPackagesAvailable(c("Rsamtools"))
+    if (!is.null(regions) && !inherits(regions, "GRanges")) {
+        stop("'regions' must be either NULL or a GRanges object.")
+    }
+    if (!is.na(nmax) && (!is.numeric(nmax) || length(nmax) != 1L || nmax <= 0)) {
+        stop("'nmax' must be either NA or a numeric scalar greater than zero.")
+    }
+    if (!is.numeric(isizemax) || length(isizemax) != 1L || isizemax <= 0) {
+        stop("'isizemax' must be a numeric scalar greater than zero.")
+    }
+    
+    # set scanning parameters
+    if (is.null(regions)) {
+        regions <- IRanges::IRangesList()
+    }
+    params <- Rsamtools::ScanBamParam(
+        flag = Rsamtools::scanBamFlag(isPaired = TRUE,
+                                      isProperPair = TRUE,
+                                      isUnmappedQuery = FALSE,
+                                      hasUnmappedMate = FALSE,
+                                      isFirstMateRead = TRUE,
+                                      isSecondaryAlignment = FALSE,
+                                      isNotPassingQualityControls = NA,
+                                      isDuplicate = FALSE,
+                                      isSupplementaryAlignment = FALSE),
+        what = c("isize"),
+        which = regions)
+    
+    # open bam file(s)
+    bfl <- Rsamtools::BamFileList(fname, yieldSize = min(2e6, nmax, na.rm = TRUE))
+    
```

Comment:
is the number of reads read capped at 2e6 here? If so, would need to be mentioned in the help function

Reply:
Sorry I misunderstood, I think `yieldSize` is just related to the number of reads you read in together at an instance, but not the final number of reads read in.

Reply:
Yes, that is correct. The while loop will iterate as many times as needed, but this will make sure that `nmax` is respected in case it is set lower than the yieldSize.